src/linux/ebpf.cpp

```diff
@@ -151,6 +163,52 @@ Try<Nothing> attach(int fd, const string& cgroup)
   return Nothing();
 }
 
+
+Try<vector<uint32_t>> attached(const string& cgroup)
+{
+  Try<int> cgroup_fd = os::open(cgroup, O_DIRECTORY | O_RDONLY | O_CLOEXEC);
+  if (cgroup_fd.isError()) {
+    return Error("Failed to open '" + cgroup + "': " + cgroup_fd.error());
+  }
+
+  // Program ids are unsigned 32-bit integers. We assume that a maximum
+  // of 32 programs are attached to a cgroup; there should only be 0 or 1 but
+  // we allow for more to be safe. We initialize our query buffer with -1 so
+  // we can distinguish between a returned program id and filler values.
+  //
+  // Example: [1, 7, 3, -1, -1, -1, ..., -1] => [1, 7, 3] are program ids.
+  const int MAX_IDS = 32;
+  vector<int32_t> buffer(MAX_IDS, -1);
+
+  bpf_attr attr;
+  memset(&attr, 0, sizeof(attr));
+  attr.query.target_fd = *cgroup_fd;
+  attr.query.attach_type = BPF_CGROUP_DEVICE;
+  attr.query.prog_cnt = MAX_IDS;
+  attr.query.prog_ids = reinterpret_cast<uint64_t>(buffer.data());
+
+  Try<int, ErrnoError> result = bpf(BPF_PROG_QUERY, &attr, sizeof(attr));
+
+  if (result.isError()) {
+    os::close(*cgroup_fd);
+    return Error(
+        "BPF syscall to find a program ids attached to '" + cgroup +
+        "' failed: " + result.error().message);
```

Comment:
the caller is going to know that we're trying to find device programs attached to the given cgroup, i.e. the caller will log something like:

```
Try<x> attached = ebpf::cgroups2::attached("/some/cgroup/path");
if error:
  return Error("Failed to find bpf programs attached to '/some/cgroup/path': " + attached.error());
```

So to compose the messages well, here we can do something like:

```
return result.error();
```

or:

```
return Error("bpf syscall to BPF_PROG_QUERY for BPF_CGROUP_DEVICE programs failed: " + result.error().message);
```

Since these ^ are the piece of information the caller doesn't know and won't include when composing.

Reply:
Went with the second one.